rudder-core/src/main/scala/com/normation/rudder/services/reports/ExecutionBatch.scala

```diff
@@ -282,15 +322,183 @@ object ExecutionBatch extends Loggable {
+   * given node config info.
+   *
+   * The method returns either an error (no expected reports, bad config id, no reports at all...)
+   * or a datastructure with the Correct state and its validity interval, and the next state after the
+   * interval (before interval does not make sense, the begining of the interval is the run execution time)
+   */
+  def computeNodeRunAndConfigInfo(
+      runInfos         : AgentRunWithNodeConfig
+    , currentNodeConfig: Option[NodeExpectedReports]
+    , nodeConfigIdInfos: Option[Seq[NodeConfigIdInfo]]
+  ): Either[RunAndConfigInfo with RunKind.Invalid, RunAndConfigInfoInterval] = {
+    implicit val _n = runInfos.agentRunId.nodeId
+
+    (runInfos, currentNodeConfig) match {
+
+      //
+      // #1 : What the hell ?
+      //      that's not good. Why a run without expected config ?
+      //      More over, we group together the cases where we have config for the run, because without
+      //      a current expected config, it should be impossible (no way to paired it with that run).
+      //      Just log an error for dev.
+      case ((AgentRunWithNodeConfig(AgentRunId(_, t), optConfigId, _, _)), None) =>
+        if(nodeConfigIdInfos.isDefined) {
+          Left(NoExpectedReport("nodeId exists in DB but has no version (due to cleaning?). Need regeneration, no expected report yet.", t, optConfigId.map( _._1)))
+        } else {
+          Left(NoExpectedReport("nodeId was not found in DB but is sending reports. It is likely a new node which needs regeneration, no expected report yet."
+          , t, optConfigId.map( _._1))
+          )
+        }
+
+      //
+      // #2 : run without config ID (neither in it nor found)
+      //      no expected config for the run. Why so? At least, a recent config.
+      case ((AgentRunWithNodeConfig(AgentRunId(_, t), None, _, _)), Some(currentConfig)) =>
+        /*
+         * Here, we want to check two things:
+         * - does the run should have contain a config id ?
+         *   It should if the oldest config was created a long time ago,
+         *   and if it is the case most likelly the node can't get
+         *   its updated promises.
+         *   The logic is that only nodes with initial promises send reports without a config Id. So
+         *   if a node is in that case, it is because it never got genererated promises.
+         *   If the first generated promises for that node are beyond the grace period, it means that
+         *   the run should have used theses promises, and we have a (DNS) problem because it didn't.
+         *
+         * - else, we look at the most recent
+         *   config and decide between pending / no answer
+         *
+         * Note: we must use value from current expected config for modes,
+         *       because we don't have anything else really.
+         */
+        val oldestConfigId = nodeConfigIdInfos.getOrElse(Seq(currentConfig.configInfo)).minBy( _.creation.getMillis)
```

Comment:
why not using the one from the nodeConfig timestamp ?

Reply:
I'm not sure about the question, I'm trying to reformulate it. So you are wondering why, when I have a run without a configId, I'm looking for the oldest configuration, and not the configuration that was generated around the run timestamp? 
The reason is that the question I'm trying to answer is: "is it legit for that node to send a run without a configId?". And it is legit ONLY if the node never updated its promises and run initial promises, and so I need to know when was the first policy generation done for that node, to see if it is ok that the node is not yet sending reports for it. 

If I didn't understood your question, don't hesitate to comment further.

Reply:
ok, you answered a question i didn't know I wanted to ask :)
So, my real question is that we use oldest config id, and report on it that we have a UnexpectedNoVersion, but really, we have a UnexpectedNoVersion with the configId of the timestamp of the run we are considering.  So if (oldestExpiration.isBefore(t)) (which is the point i was missing, and thank you for your answer) the you should Left(UnexpectedNoVersion(...  with the configId of the time of the run, so that we know what was expected on the node at this time, and what was missing

Reply:
That makes sense